### ISSUE TITLE
Multiple versions, with shuffled answers, of the same questionnaire

### ISSUE THREAD

**Issue by Miguel-Hermanns** [sdaps/sdaps#239]
Is there a way to use sdaps to generate multiple versions, with multiple-choice answers being shuffled, of the same questionnaire? I'm thinking of the application of sdaps to generate exams. 

A clumpy way of doing it, of course, it to generate by hand different questionnaires, with different QR codes, and process each version separately. But a more streamlined way of doing it would be great.

Thanks in advance and best regards,

Miguel

**Comment by benzea** [comment]
Unfortunately no.

You could script something, but at the end of the day you need separate SDAPS projects for this. One method that some people have used is to just have a generic answer sheet, and store the information separately. Obviously, that only works for true multiple choice tests.

We could probably improve it, and if it is just by having a multi-project mode in SDAPS, which just allows throwing in a scan and selecting the correct "survey" (internally). There isn't really anything that prevents us from doing that from a technical perspective, as long the main parameters (paper size, margins, etc. match). Reason being, that we already have an sqlite database that allows storing multiple "survey"s in one project directory.

But, for now, you are stuck with multiple projects. Then, just throw the scan into all of these projects, rely on SDAPS marking the rows as invalid and merging the data. Or, pre-sort the images by reading the QR code yourself.

**Comment by Miguel-Hermanns** [comment]
Thanks for the fast response. The problem of doing it by hand, or with an external script, is that it is very error prone. That is the reason why I want to try to avoid it if possible. At the moment I generate the versions automatically, using the shuffle functions of the pgf package. I also generate automatically the corresponding correction sheets in order to minimise possible typos. And it works perfectly.

Another idea. Imagine I want to generate ten versions of a multiple choice questionnaire. Could I generate a ten pages questionnaire, with each page containing one version of the true questionnaire? Each student would then fill out one page only, and when processing the questionnaire each student would have nine pages missing and one page correctly read. Does it make sense? Would it be feasible?

**Comment by Miguel-Hermanns** [comment]
Your first point is an issue, I think. If everybody is marked as invalid, how do I know who has double mark issues, etc? Or is a different error code being set in the exported csv in that case?

Regarding your second comment, when you say "by adding the images individually" what do you mean exactly? I will not be able to feed sdaps with a 600 pages pdf file directly and instead I need to feed it with 600 individual files?

**Comment by benzea** [comment]
> Your first point is an issue, I think. If everybody is marked as invalid, how do I know who has double mark issues, etc? Or is a different error code being set in the exported csv in that case?

Well, assuming that something checked is detect on the page, you'll know which of the questions have answers. For these, you'll get the correct checkbox information (and for single choice, you get an error code if two or none are selected).

There is also the `valid` entry in the CSV, that will be false.

> Regarding your second comment, when you say "by adding the images individually" what do you mean exactly? I will not be able to feed sdaps with a 600 pages pdf file directly and instead I need to feed it with 600 individual files?

Yes, exactly that.

**Comment by benzea** [comment]
On the first point. It is obviously possible to change the exported data a bit to contain more information.

**Comment by benzea** [comment]
@Miguel-Hermanns are you planning to try something out or did you give up for now on using SDAPS for this purpose?

Just curious, because if you have some (experimental) code, I would be willing to take a look. And, if it works, it might be nice to create an example out of it for others.

**Comment by Miguel-Hermanns** [comment]
I want to try something out, but not now. Probably sometime in Spring. I will let you know. I need to figure out, however, how to automate the correction afterwards. For it, the LaTeX code generating the different versions needs to export something that can then be imported into Excel together with the csv file generated by SDAPS. As said, I still need to think on it.

**Comment by BjornFJohansson** [comment]
I found this conversation very interesting. Although this is off-topic, anyone interested should have a look at [AMC](https://www.auto-multiple-choice.net) a set of perl scripts that does multiple choice exams.

**Comment by benzea** [comment]
Yup, AMC is designed to do this. It is even listed at the bottom of https://sdaps.org/

Another way would be for SDAPS to get a multi-survey mode. Generate all combinations, give each a unique QR code, insert into the same DB and just make sure that some properties (like page count) is the same. Then, just select the correct survey in the DB each time ...

Anyway, I guess I can leave this open for now. Maybe an FAQ or so somewhere would be good.

**Comment by Miguel-Hermanns** [comment]
I think the solution you propose with the multi-survey mode should work in my case as everything has the same number of pages, namely, two. I could scan together absolutely everything ("white sheets" with the student's developments and the multiple versions of the questionnaires) and let SDAPS deal with it. I would only need to worry then with the post-processing of everything, for which I would probably stop using Excel and switch to Python or something else.
